File: src/hvm/Calculadora.cpp
#include "hvm/Calculadora.hpp"
#include <iostream>
#include <stdexcept>

void Calculadora::soma(int valor) {
    if (valor < 0 || valor + acumulador > 999) {
        throw std::runtime_error("Erro na operação " + std::to_string(acumulador) + " + " + std::to_string(valor) + " = " + std::to_string(valor + acumulador) + ", único valor aceito como resultado é entre 0-999");
    }

    acumulador += valor;
}

void Calculadora::subtraia(int valor) {
    if (valor < 0 || valor - acumulador > 999) {
        throw std::runtime_error("Erro na operação " + std::to_string(acumulador) + " - " + std::to_string(valor) + " = " + std::to_string(valor - acumulador) + ", único valor aceito como resultado é entre 0-999");
    }

    acumulador -= valor;
}

void Calculadora::multiplicar(int valor) {
    if (valor < 0 || valor * acumulador > 999) {
        throw std::runtime_error("Erro na operação " + std::to_string(acumulador) + " * " + std::to_string(valor) + " = " + std::to_string(valor * acumulador) + ", único valor aceito como resultado é entre 0-999");
    }

    acumulador *= valor;
}

void Calculadora::divida(int valor) {
    
    int integer_value = static_cast<int>(acumulador / valor);

    if (valor < 0 || integer_value > 999) {
        throw std::runtime_error("Erro na operação " + std::to_string(acumulador) + " / " + std::to_string(valor) + " = " + std::to_string(integer_value) + ", único valor aceito como resultado é entre 0-999");
    }
    
    acumulador = integer_value; 
}

void Calculadora::acumular(int valor) {
    if (valor < 0 || valor > 999) {
        throw std::runtime_error("Erro na escrita do valor " + std::to_string(valor) + " no acumulador. Apenas valores entre 0-999 são aceitos.");
    }

    acumulador = valor;
}

int Calculadora::getAcumulador() {
    return acumulador;
}

File: src/public/ts/hvm/Calculadora.ts
import HVC from "../HVC";

export default class Calculadora {
    
    private acumulador = 0;
  
    public soma(valor:number) {
      
      if (valor + this.acumulador > 999) {
        HVC.terminal.addError(`Erro na operação ${this.acumulador} + ${valor} = ${valor + this.acumulador}, único valor aceito como resultado é entre 0-999`);
        return "erro"
      }

      this.acumulador += valor;

      return "sucesso"
      
    }
  
    public subtraia(valor:number) {
      
      if (this.acumulador - valor > 999) {
        HVC.terminal.addError(`Erro na operação ${this.acumulador} - ${valor} = ${valor - this.acumulador}, único valor aceito como resultado é entre 0-999`);
        return "erro"
      }
      this.acumulador -= valor;
      return "sucesso"
    }
  
    public multiplicar(valor:number) {
      
      if (valor * this.acumulador > 999) {
        HVC.terminal.addError(`Erro na operação ${this.acumulador} * ${valor} = ${valor * this.acumulador}, único valor aceito como resultado é entre 0-999`);
        return "erro"
      }
      this.acumulador *= valor;
      return "sucesso"
    }
  
    public divida(valor:number) {

      if (valor < 0 || valor / this.acumulador > 999) {
        HVC.terminal.addError(`Erro na operação ${this.acumulador} / ${valor} = ${valor / this.acumulador}, único valor aceito como resultado é entre 0-999`);
        return "erro"
      }
      // this.acumulador = parseInt((parseInt(this.acumulador.toString()) / parseInt(valor.toString())).toString());

      this.acumulador = Math.floor(this.acumulador / valor)
      return "sucesso"

    }

    public acumular(valor:number) {
    
        if(valor < 0 || valor > 999) {
    
            HVC.terminal.addError(`Erro na escrita do valor ${valor} no gaveteiro. Apenas valores entre 0-999 são aceitos.`);
            return "erro"

        }
    
        this.acumulador = valor;

        return "sucesso"
    
    }
    
    public getAcumulador() {
    
      return this.acumulador;
    
    }

}